syntax = "proto3";
package music_catalog;

import "google/protobuf/timestamp.proto";

message Artist {
  int32 artist_id = 1;
  string name = 2;
  bool is_verified = 3;
}

message Genre {
  int32 genre_id = 1;
  string name = 2;
}

message Track {
  int32 track_id = 1;
  string title = 2;
  repeated Artist artists = 3;
  repeated Genre genres = 4;
  int32 duration_ms = 5;
  bool explicit = 6;
  string release_date = 7;
  google.protobuf.Timestamp created_at = 8;
}

message Pagination {
  int32 offset = 1;
  int32 limit = 2;
  int32 total = 3;
}

message GetTracksByArtistRequest {
  int32 artist_id = 1;
  Pagination pagination = 2;
}

message GetTracksByGenreRequest {
  int32 genre_id = 1;
  Pagination pagination = 2;
}

message TrackListResponse {
  repeated Track tracks = 1;
  Pagination pagination = 2;
}

message VerifyTrackRequest {
  string track_id = 1;
}

message VerifyTrackResponse {
  bool exists = 1;
}

message GetTrackRequest {
  int32 track_id = 1;
}

service TrackQueryService {
  rpc GetTracksByArtist(GetTracksByArtistRequest) returns (TrackListResponse);
  rpc GetTracksByGenre(GetTracksByGenreRequest) returns (TrackListResponse);
}

// service TrackService {
//   rpc VerifyTrackExists(VerifyTrackRequest) returns (VerifyTrackResponse);
//   rpc GetTrack(GetTrackRequest) returns (Track);
// }